runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.nonrepudiation
+
+import java.security.{MessageDigest, PublicKey}
+
+import com.google.common.io.BaseEncoding
+
+import scala.collection.concurrent.TrieMap
+
+object KeyRepository {
+
+  trait Read {
+    def get(fingerprint: String): Option[PublicKey]
+  }
+
+  trait Write {
+    def put(key: PublicKey): String
+  }
+
+  private def hash(key: PublicKey): String = {
+    val sha256: MessageDigest = MessageDigest.getInstance("SHA-256")
+    val fingerprint: Array[Byte] = sha256.digest(key.getEncoded)
+    BaseEncoding.base64.encode(fingerprint)
+  }
+
+  final class InMemory(keys: PublicKey*) extends KeyRepository {
+
+    private val map: TrieMap[String, PublicKey] = TrieMap(keys.map(key => hash(key) -> key): _*)
```

Comment:
maybe newtype the fingerprint?

Reply:
You mean wrapping the `String` in a more specific class to make it more type safe, right? Sure, why not.

Reply:
Yeah, wrap it in a case class or `scalaz` Tag magic which has no runtime overhead.

Reply:
Addressed in https://github.com/digital-asset/daml/pull/8608/commits/70c21416b7570b02174ad0481e7c512989b61b6d

For know I went with the same approach we use in Daml-LF, with a subtype of `String`. I'll switch to a `Tag` if I start using more `scalaz` stuff.